### PR TITLE
feat(curriculum): add css colors lab

### PR DESCRIPTION
**Pull request by ihechikara**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Comment by moT01** [comment]
A PR has been merged that renamed the certification from "Front End" to "Full Stack". This PR may need an update and a few changes made: [see comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/56694#issuecomment-2444473378)

**Comment by moT01** [comment]
A PR has been merged that moves the block order from the `meta.json` file to the `full-stack.json` file.](https://github.com/freeCodeCamp/freeCodeCamp/pull/56795) This PR will need an update, and the order will need to be removed from the `meta.json` file and moved to the new file.

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
+---
+id: 66f3f6eb66ea9dc41cdc30df
+title: Design Colored Boxes
+challengeType: 14
+dashedName: "lab-colored-boxes"
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:5]:
```suggestion
dashedName: lab-colored-boxes
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+const height = colorBoxStyles.height;
+
+assert.notStrictEqual(width, '0px');
+assert.notStrictEqual(height, '0px');
+```
+
+The `.color1` element should have a hexadecimal background color.
+
+```js
+assert.match(code, /\.color1\s*{[^}]*\bbackground-color\s*:\s*#[0-9a-fA-F]{3,6}\s*;[^}]*}/);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:82]:
```suggestion
assert.match(
  new __helpers.CSSHelp(document).getStyle('.color1').getPropVal('background-color', true),
  /^#[0-9a-fA-F]{3,6}$/
);
```

wouldn't a test like this avoid the issue of comments, and eventual other properties?

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+The `.color2` element should have an RGB background color.
+
+```js
+assert.match(code, /\.color2\s*{[^}]*\bbackground-color\s*:\s*rgb\s*\(\s*\d+\s*,\s*\d+\s*,\s*\d+\s*\)\s*;[^}]*}/);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:88]:
```suggestion
assert.match(
  new __helpers.CSSHelp(document).getStyle('.color2').getPropVal('background-color', true),
  /^rgb\s*\(\s*\d+\s*,\s*\d+\s*,\s*\d+\s*\)$/
);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+The `.color3` element should have a predefined (word) background color.
+
+```js
+assert.match(code, /\.color3\s*{[^}]*\bbackground-color\s*:\s*[a-zA-Z]+\s*;[^}]*}/);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:94]:
```suggestion
assert.match(
  new __helpers.CSSHelp(document).getStyle('.color3').getPropVal('background-color', true),
  /^[a-zA-Z]+$/
);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+The `.color4` element should have a HSL background color.
+
+```js
+assert.match(code, /\.color4\s*{[^}]*\bbackground-color\s*:\s*hsl\s*\(\s*\d+\s*,\s*\d+%\s*,\s*\d+%\s*\)\s*;[^}]*}/);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:100]:
```suggestion
assert.match(
  new __helpers.CSSHelp(document).getStyle('.color4').getPropVal('background-color', true),
  /^hsl\(\d+,\d{1,3}%,\d{1,3}%\)$/
);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+The `.color5` element should have a background color set.
+
+```js
+assert.match(code, /\.color5\s*{[^}]*\bbackground-color\s*:\s*[^;]+;[^}]*}/);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:106]:
```suggestion
assert.isNotEmpty(
  new __helpers.CSSHelp(document).getStyle('.color5').getPropVal('background-color', true)
);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 2 above)
Comment (ihechikara) [reply curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:82]: @ilenia-magoni, the issue with these formats is how the browser handles/converts different CSS rules. For example:

```js
assert.match(
  new __helpers.CSSHelp(document).getStyle('.color2').getPropVal('background-color', true),
  /^rgb\s*\(\s*\d+\s*,\s*\d+\s*,\s*\d+\s*\)$/
);
```

The test above will pass for both rgb and hsl. The tests you suggested for `.color1` and `.color4` don't work, but with a little workaround, `.color1` starts passing for hex, rgb, and hsl values.

IMO, using regex here ensures each test passes for specific rules. I couldn't find a workaround before switching to regex.

CC @moT01 @jdwilkin4

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 6 above)
Comment (ilenia-magoni) [reply curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:106]: This is only checking if there is a value for the background-color not which value it is, at least this one is fine

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 2 above)
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:82]:
```suggestion
assert.match(__helpers.removeCssComments(code), /\.color1\s*{[^}]*\bbackground-color\s*:\s*#[0-9a-fA-F]{3,6}\s*;[^}]*}/);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 3 above)
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:88]:
```suggestion
assert.match(__helpers.removeCssComments(code), /\.color2\s*{[^}]*\bbackground-color\s*:\s*rgb\s*\(\s*\d+\s*,\s*\d+\s*,\s*\d+\s*\)\s*;[^}]*}/);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 4 above)
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:94]:
```suggestion
assert.match(__helpers.removeCssComments(code), /\.color3\s*{[^}]*\bbackground-color\s*:\s*[a-zA-Z]+\s*;[^}]*}/);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 5 above)
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:100]:
```suggestion
assert.match(__helpers.removeCssComments(code), /\.color4\s*{[^}]*\bbackground-color\s*:\s*hsl\s*\(\s*\d+\s*,\s*\d+%\s*,\s*\d+%\s*\)\s*;[^}]*}/);
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 1 above)
Comment (ilenia-magoni) [reply curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:5]: no other challenge has quotes there

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+2. You should have a `div` with a class of `color-grid` to hold all your color elements.
+3. You should have five `div` elements within `color-grid` `div`.
+4. The five `div` elements should each have a class of `color-box` and `color#`, where `#` is the number of the order of that `div`. For example: `color1` for the first `div`, `color2` for the second, and so on.
+5. The `.color-box` class should have a set `width` and `height` so your `div` elements are visible on the page.
+6. The `.color1` element should have a `background-color` that uses hexadecimal color value.
+7. The `.color2` element should have a `background-color` that uses an RGB color value.
+8. The `.color3` element should have a `background-color` that uses a predefined (word) color value.
+9. The `.color4` element should have a `background-color` that uses a HSL color value.
+10. The `.color5` element should have a `background-color` set.
+
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:27]:
```suggestion

**Note:** Be sure to link your stylesheet in your HTML and apply your CSS.

```

**Review by moT01 — COMMENTED** [review]
This is looking good @ihechikara 👍 

I left one suggestion to add the note about linking the stylesheet. Also, I think we should rename this to `Design a Set of Colored Boxes`.

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+In this lab, you'll practice using CSS colors by designing boxes.
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.
+
+**User Stories:**
+
+1. You should set the background color for `body` to `#f4f4f4`.
+2. You should have a `div` with a class of `color-grid` to hold all your color elements.
+3. You should have five `div` elements within `color-grid` `div`.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:19]:
```suggestion
3. You should have five `div` elements within `div.color-grid`.
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+```
+
+You should have a `div` element with a class of `color-grid`.
+
+```js
+const colorGrid = document.querySelector('div.color-grid');
+assert.exists(colorGrid);
+```
+
+You should have five `div` elements within the `.color-grid` `div` element.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:47]:
```suggestion
You should have five `div` elements within the `div.color-grid` element.
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 8 above)
Comment (moT01) [reply curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:19]: I personally would use ``You should have five `div` elements within the `.color-grid` element.``

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md (hunk 9 above)
Comment (moT01) [reply curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:47]: same here

File: curriculum/challenges/_meta/lab-colored-boxes/meta.json [hunk 10]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "Design a Set of Colored Boxes",
+  "isUpcomingChange": true,
+  "usesMultifileEditor": true,
+  "dashedName": "lab-colored-boxes",
+  "superBlock": "full-stack-developer",
+  "challengeOrder": [{ "id": "66f3f6eb66ea9dc41cdc30df", "title": "Design Colored Boxes" }],
```
Comment (moT01) [inline curriculum/challenges/_meta/lab-colored-boxes/meta.json:7]:
```suggestion
  "challengeOrder": [{ "id": "66f3f6eb66ea9dc41cdc30df", "title": "Design a Set of Colored Boxes" }],
```

File: curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
+---
+id: 66f3f6eb66ea9dc41cdc30df
+title: Design a Set of Colored Boxes
+challengeType: 14
+dashedName: lab-colored-boxes
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-colored-boxes/66f3f6eb66ea9dc41cdc30df.md:5]:
```suggestion
dashedName: set-of-colored-boxes
```

A little confusing, but this seems to be the standard.

**Comment by moT01** [comment]
I think we want to change the challengeType to 25 here. [See PR](https://github.com/freeCodeCamp/freeCodeCamp/pull/57441)

**Comment by ihechikara** [comment]
> I think we want to change the challengeType to 25 here. [See PR](https://github.com/freeCodeCamp/freeCodeCamp/pull/57441)

Done

**Review by moT01 — APPROVED** [review]
LGTM 🎉 

Note - to test this PR at the moment, you either need to change the challengeType back to 14 or update the branch and add the changes from [github.com/freeCodeCamp/freeCodeCamp/pull/57507](https://github.com/freeCodeCamp/freeCodeCamp/pull/57507). We are fine to merge it as-is.

**Review by Sembauke — APPROVED** [review]

**Pull request merged**